step-8-edge-cases.md

```diff
@@ -50,22 +52,54 @@ But since we're not using Spring Context here, we need to create an instance of
 
 The only missing part is `LettuceConnectionFactory`'s arguments, Redis' host and port.
 
-We will use a JUnit Rule for starting Redis:
+We will use Testcontainers' JUnit Jupiter extension for starting Redis:
 
 ```java
-    @Rule
+    @Container
     public GenericContainer redis = new GenericContainer("redis:3-alpine")
             .withExposedPorts(6379);
 ```
 
-And set it:
-
+And add the `@Testcontainers` annotation to the class: 
+```java
+@Testcontainers
+public class RatingsRepositoryTest {
+```
+And the code for initializing the connection factory:
 ```java
 LettuceConnectionFactory connectionFactory = new LettuceConnectionFactory(
-        redis.getContainerIpAddress(),
+        redis.host(),
         redis.getFirstMappedPort()
 );
 ```
 
-If we run the tests, we discover that there is a limit of 64bit number for the increment, oh well!
+The test should fail with a somewhat cryptic error:
+```text
+Error in execution; nested exception is io.lettuce.core.RedisCommandExecutionException: ERR increment or decrement would overflow
+```
+And there's nothing on your side to fix, the test is pushing the boundaries of Redis.  
+But on the bright side we learned how to use Testcontainers outside of the Spring Framework. 
+We also saw how we can utilize to learn about the limitations and behavior of extra components.
+
+Delete the test before anyone notices. Just kidding, let's turn this into a useful test by asserting we get the exceptions.
```

Comment:
IMO we shouldn't be asserting faulty behaviour (nobody would expect a request to fail on production, so why would test do?)

Reply:
The alternative (I think we discussed with @rnorth) would be to change the production code to validate the input and consciously throw an `InvalidArgumentException`, which we could then assert here.

Reply:
That's one of the reasons why this step never had a solution offered - the test shouldn't be changed (this would be a really bad thing to recommend doing, I think we all agree here), and it is unclear how to change the implementation (validate? Catch and ignore the overflow? Something else?)

Anyways, if we want to provide a solution, I suggest we change the implementation and not the test. After all, this isn't test's fault :D

Reply:
IMO the test helped us in exploring the behavior of the product we are using. With the behavior being discovered through the test, it is IMO very valid to change the implementation and maybe even the test itself to incorporate the change in design (this is an interactive feedback loop, in the style of TDD). 

So I think it would make sense to change the prod code to throw a meaningful custom exception (`MaxRatingsAddedException`?) and assert this in the test. I don't think the repository can make any meaningful business decisions about what else to do (catch and ignore feels weird).

Reply:
Okay, as a compromise, `MaxRatingsAddedException` would be fine (and pull the decision on how to handle it on the upper level).

I am still not in favour of changing a test here (because a good test should reflect how system should work, not how it is implemented), but `MaxRatingsAddedException` would at least remove implementation-specific details (`RedisSystemException` is cryptic)

Reply:
Sounds good 👍 

> I am still not in favour of changing a test here (because a good test should reflect how system should work, not how it is implemented), but MaxRatingsAddedException would at least remove implementation-specific details

Generally yes, but sometimes you learn about the limitations of the system only during implementation and then you might need to make new decisions on how the business logic needs to handle this (or if this limitation is mostly a theoretical one).

Whatever, I think this discussion is somewhat academic and not that important for this small example, but I think it's great we have it. I will adapt it accordingly in the docs 👍